File: app/src/main/java/dustit/clientapp/mvp/ui/fragments/FeedFragment.java
package dustit.clientapp.mvp.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import dustit.clientapp.R;
import dustit.clientapp.customviews.WrapperLinearLayoutManager;
import dustit.clientapp.mvp.model.entities.FavoriteEntity;
import dustit.clientapp.mvp.model.entities.MemEntity;
import dustit.clientapp.mvp.presenters.fragments.FeedFragmentPresenter;
import dustit.clientapp.mvp.ui.adapters.FeedRecyclerViewAdapter;
import dustit.clientapp.mvp.ui.base.BaseFeedFragment;
import dustit.clientapp.mvp.ui.interfaces.IFeedFragmentView;
import dustit.clientapp.utils.AlertBuilder;
import dustit.clientapp.utils.L;
import dustit.clientapp.utils.managers.ThemeManager;

import static android.support.v7.widget.RecyclerView.SCROLL_STATE_IDLE;


public class FeedFragment extends BaseFeedFragment implements IFeedFragmentView, FeedRecyclerViewAdapter.IFeedInteractionListener {

    public static final String HEIGHT_APPBAR = "HEIGHT";
    private int appBarHeight;

    @BindView(R.id.rvFeed)
    RecyclerView rvFeed;
    @BindView(R.id.srlFeedRefresh)
    SwipeRefreshLayout srlRefresh;

    private Unbinder unbinder;

    private FeedRecyclerViewAdapter adapter;

    private FeedFragmentPresenter presenter;

    private RecyclerView.OnScrollListener scrollListener;

    private WrapperLinearLayoutManager linearLayoutManager;

    @Inject
    ThemeManager themeManager;

    public FeedFragment() {
        // Required empty public constructor
    }

    public void setFavoritesList(List<FavoriteEntity> list) {
        adapter.setFavoritesList(list);
    }

    public static FeedFragment newInstance(int appBarHeight) {
        Bundle args = new Bundle();
        args.putInt(HEIGHT_APPBAR, appBarHeight);
        final FeedFragment fragment = new FeedFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void setArguments(@Nullable Bundle args) {
        super.setArguments(args);
        if (args != null) {
            appBarHeight = args.getInt(HEIGHT_APPBAR);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        bindWithBase(context);
        L.print("onAttach");
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_feed, container, false);
        unbinder = ButterKnife.bind(this, v);
        linearLayoutManager = new WrapperLinearLayoutManager(getContext());
        rvFeed.setLayoutManager(linearLayoutManager);
        adapter = new FeedRecyclerViewAdapter(getContext(), this, appBarHeight);
        rvFeed.setAdapter(adapter);
        presenter = new FeedFragmentPresenter();
        presenter.bind(this);
        srlRefresh.setProgressViewOffset(false, appBarHeight, appBarHeight + 100);
        presenter.loadBase();
        srlRefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                srlRefresh.setRefreshing(true);
                presenter.loadBase();
            }
        });
        scrollListener = new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                if (newState == SCROLL_STATE_IDLE) {
                    notifyFeedScrollIdle(true);
                    if (rvFeed.getChildAt(0) != null)
                        if (rvFeed.getChildAt(0).getTop() == appBarHeight && linearLayoutManager.findFirstVisibleItemPosition() == 0) {
                            if (!srlRefresh.isRefreshing())
                                notifyFeedOnTop();
                        }
                } else {
                    notifyFeedScrollIdle(false);
                }
            }

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                notifyFeedScrollChanged(dy);
            }
        };
        rvFeed.addOnScrollListener(scrollListener);
        return v;
    }

    @Override
    public void onDestroyView() {
        rvFeed.removeOnScrollListener(scrollListener);
        unbinder.unbind();
        presenter.unbind();
        super.onDestroyView();
    }

    @Override
    public void onDetach() {
        super.onDetach();
    }

    @Override
    public void onBaseUpdated(List<MemEntity> list) {
        srlRefresh.setRefreshing(false);
        adapter.updateListWhole(list);
    }

    @Override
    public void onPartialUpdate(List<MemEntity> list) {
        adapter.updateListAtEnding(list);
    }

    @Override
    public void onErrorInLoading() {
        srlRefresh.setRefreshing(false);
        adapter.onFailedToLoad();
    }

    @Override
    public void onStartLoading() {
        adapter.onStartLoading();
    }

    @Override
    public void onLikePostError(String id) {
        showErrorToast();
    }

    @Override
    public void onLikeDeletingError(String id) {
        showErrorToast();
    }

    @Override
    public void onDislikePostError(String id) {
        showErrorToast();
    }

    @Override
    public void onDislikeDeletingError(String id) {
        showErrorToast();
    }

    @Override
    public void onLikePostedSuccessfully(String id) {
        adapter.onLikePostedSuccesfully(id);
    }

    @Override
    public void onLikeDeletedSuccessfully(String id) {
        adapter.onLikeDeletedSuccesfully(id);
    }

    @Override
    public void onDislikePostedSuccessfully(String id) {
        adapter.onDislikePostedSuccesfully(id);
    }

    @Override
    public void onDislikeDeletedSuccessfully(String id) {
        adapter.onDislikeDeletedSuccesfully(id);
    }

    @Override
    public void onAddedToFavorites(String id) {
        notifyBase(id);
        adapter.addedToFavorites(id);
    }

    @Override
    public void onErrorInAddingToFavorites(String id) {
        showErrorToast();
    }

    @Override
    public void onErrorInRemovingFromFavorites(String s) {
        showErrorToast();
    }

    @Override
    public void onRemovedFromFavorites(String s) {
        adapter.onDeletedFromFavorites(s);
    }

    @Override
    public void reloadFeedPartial(int offset) {
        presenter.loadWithOffset(offset);
    }

    @Override
    public void reloadFeedBase() {
        presenter.loadBase();
    }

    @Override
    public void onMemSelected(View view, MemEntity mem) {
        launchMemView(view, mem);
    }

    @Override
    public void postLike(String id) {
        presenter.postLike(id);
    }

    @Override
    public void deleteLike(String id) {
        presenter.deleteLike(id);
    }

    @Override
    public void postDislike(String id) {
        presenter.postDislike(id);
    }

    @Override
    public void deleteDislike(String id) {
        presenter.deleteDislike(id);
    }

    @Override
    public void addToFavorites(String id) {
        presenter.addToFavorites(id);
    }

    @Override
    public void deleteFromFavorites(String id) {
        presenter.removeFromFavorites(id);
    }

    @Override
    public void showErrorToast() {
        Toast.makeText(getContext(), getString(R.string.error), Toast.LENGTH_SHORT).show();
    }

    public void passPostLike(String id) {
        adapter.onLikePostedSuccesfully(id);
    }

    public void passDeleteLike(String id) {
        adapter.onLikeDeletedSuccesfully(id);
    }

    public void passPostDislike(String id) {
        adapter.onDislikePostedSuccesfully(id);
    }

    public void passDeleteDislike(String id) {
        adapter.onDislikeDeletedSuccesfully(id);
    }
    @Override
    public void onNotRegistered() {
        AlertBuilder.showNotRegisteredPrompt(getContext());
    }
}

File: app/src/main/java/dustit/clientapp/mvp/ui/fragments/FeedFragment.kt
package dustit.clientapp.mvp.ui.fragments

import android.app.ActivityOptions
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.SimpleItemAnimator
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import butterknife.ButterKnife
import butterknife.Unbinder
import dustit.clientapp.R
import dustit.clientapp.customviews.WrapperLinearLayoutManager
import dustit.clientapp.mvp.model.entities.MemEntity
import dustit.clientapp.mvp.model.entities.NewAchievementEntity
import dustit.clientapp.mvp.presenters.fragments.FeedFragmentPresenter
import dustit.clientapp.mvp.ui.activities.AccountActivity
import dustit.clientapp.mvp.ui.activities.PersonalSettingsActivity
import dustit.clientapp.mvp.ui.adapters.FeedRecyclerViewAdapter
import dustit.clientapp.mvp.ui.base.BaseFeedFragment
import dustit.clientapp.mvp.ui.dialog.AchievementUnlockedDialog
import dustit.clientapp.mvp.ui.interfaces.IFeedFragmentView
import dustit.clientapp.utils.AlertBuilder
import dustit.clientapp.utils.IConstants
import kotlinx.android.synthetic.main.fragment_feed.view.*

class FeedFragment : BaseFeedFragment(), IFeedFragmentView, FeedRecyclerViewAdapter.IFeedInteractionListener {
    private var appBarHeight: Int = 0
    private var rvFeed: RecyclerView? = null
    private var srlRefresh: SwipeRefreshLayout? = null
    private var unbinder: Unbinder? = null
    private var presenter: FeedFragmentPresenter? = null
    private var scrollListener: RecyclerView.OnScrollListener? = null
    private var linearLayoutManager: WrapperLinearLayoutManager? = null
    private var rlEmptyCategories: ViewGroup? = null
    private var changingCategories = false
    private var myId: String? = ""

    override fun setArguments(args: Bundle?) {
        super.setArguments(args)
        if (args != null) {
            appBarHeight = args.getInt(HEIGHT_APPBAR)
            myId = args.getString(IConstants.IBundle.MY_ID)
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        bindWithBase(context)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.fragment_feed, container, false)
        rvFeed = v.rvFeed
        srlRefresh = v.srlFeedRefresh
        unbinder = ButterKnife.bind(this, v)
        bindFeedback(this)
        linearLayoutManager = WrapperLinearLayoutManager(context)
        rvFeed!!.layoutManager = linearLayoutManager
        adapter = context?.let { FeedRecyclerViewAdapter(context, this, appBarHeight, rvFeed) }
        adapter.setHasStableIds(true)
        rvFeed!!.adapter = adapter
        rlEmptyCategories = v.hotEmpty
        setFeedPool(rvFeed!!.recycledViewPool)
        presenter = FeedFragmentPresenter()
        presenter!!.bind(this)
        v.btnEmptyHot.setOnClickListener {
            changingCategories = true
            val intent = Intent(context, PersonalSettingsActivity::class.java)
            startActivity(intent)
        }
        srlRefresh!!.setProgressViewOffset(false, appBarHeight - 100, appBarHeight + 100)
        srlRefresh!!.setOnRefreshListener {
            srlRefresh!!.isRefreshing = true
            presenter!!.loadBase()
        }
        (rvFeed!!.itemAnimator as SimpleItemAnimator).supportsChangeAnimations = false

        subscribeToFeedbackChanges()
        presenter!!.loadBase()
        return v
    }

    fun scrollToTop() {
        rvFeed?.scrollToPosition(0)
    }

    override fun isRegistered(): Boolean {
        return isUserRegistered
    }

    override fun onDestroyView() {
        unsubscribeFromFeedbackChanges()
        rvFeed!!.removeOnScrollListener(scrollListener)
        unbinder!!.unbind()
        presenter!!.unbind()
        super.onDestroyView()
    }

    override fun onResume() {
        super.onResume()
        if (changingCategories) {
            changingCategories = false
            presenter!!.loadBase()
            srlRefresh!!.visibility = View.VISIBLE
            rlEmptyCategories!!.visibility = View.INVISIBLE
        }
    }

    override fun onBaseUpdated(list: List<MemEntity>) {
        srlRefresh!!.isRefreshing = false
        adapter!!.updateWhole(list)
        rvFeed!!.scheduleLayoutAnimation()
    }

    override fun onPartialUpdate(list: List<MemEntity>) {
        if (list.isEmpty()) {
            adapter!!.onMemesEnded()
            return
        }
        adapter!!.updateAtEnding(list)
    }

    override fun onErrorInLoading() {
        srlRefresh!!.isRefreshing = false
        adapter!!.onFailedToLoad()
    }

    override fun reloadFeedBase() {
        presenter!!.loadBase()
    }

    override fun showErrorToast() {
        Toast.makeText(context, getString(R.string.error), Toast.LENGTH_SHORT).show()
    }

    override fun onNoCategories() {
        srlRefresh!!.visibility = View.INVISIBLE
        rlEmptyCategories!!.visibility = View.VISIBLE
    }

    override fun onNotRegistered() {
        AlertBuilder.showNotRegisteredPrompt(context)
    }

    override fun loadMore(offset: Int) {
        presenter!!.loadWithOffset(offset)
    }

    override fun gotoHot() {
        gotoFragment(1)
    }

    override fun onAchievementUpdate(achievementEntity: NewAchievementEntity) {
        if (context != null) {
            if (presenter!!.isRegistered)
            AchievementUnlockedDialog(context, achievementEntity.isFinalLevel).bind(achievementEntity).show()
        }
    }

    override fun gotoAccount(mem: MemEntity?) {
        if (mem != null) {
            val intent = Intent(context, AccountActivity::class.java)
            intent.putExtra(IConstants.IBundle.IS_ME, myId.equals(mem.userId))
            intent.putExtra(IConstants.IBundle.USER_ID, mem.userId)
            startActivity(intent, ActivityOptions.makeSceneTransitionAnimation(activity).toBundle())
        }
    }

    companion object {
        const val HEIGHT_APPBAR = "HEIGHT"
        fun newInstance(appBarHeight: Int, myId: String): FeedFragment {
            val args = Bundle()
            args.putInt(HEIGHT_APPBAR, appBarHeight)
            args.putString(IConstants.IBundle.MY_ID, myId)
            val fragment = FeedFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
